src/main/java/com/hubspot/jinjava/util/EagerReconstructionUtils.java

```diff
@@ -747,7 +750,14 @@ public static Set<String> resetSpeculativeBindings(
   ) {
     result
       .getSpeculativeBindings()
-      .forEach((k, v) -> replace(interpreter.getContext(), k, v));
+      .forEach(
+        (k, v) -> {
+          if (v instanceof DeferredValue) {
+            v = ((DeferredValue) v).getOriginalValue();
+          }
+          replace(interpreter.getContext(), k, v);
+        }
+      );
```

Comment:
Purpose: this is used to reset the state of the context to what it was before some logic was run, which we aren't committing.
For example if evaluating something within a deferred `{% if bool %}` block, we'll want to reset the bindings to what they used to be, this includes makes DeferredValues no longer be deferred so that the `{% else %}` block will evaluate properly. We'll then re-defer any values that got deferred during any if-elif-else blocks.
That logic already happens, I'm just explaining why we'd want to put the `DeferredValue#getOriginalValue()` onto the context here